ellie/monopoly.py

```diff
@@ -0,0 +1,154 @@
+    def __repr__(self):
+        return f"This player, {self.name}, is playing as the {self.token}. \n{self.name} has ${self.money}, is currently on {board[self.position]}, and owns these properties: {self.properties}"
+
+    def roll_dice(self):
+        # Roll two dice and return the results
+        die1 = randint(1,6)
+        die2 = randint(1,6)
+        result = die1 + die2
+        print(f"{self.name} rolled a {result} with a {die1} and a {die2}!")
+        return [die1, die2, result]
+    
+    def update_position(self):
+        # Check the dice roll and update position on the board.
+        result = self.roll_dice()
+        while result[2] > 0:
+            self.position += 1
+            result[2] -= 1
+            # Wrap around board
+            if self.position >= 40:
+                self.position = 0
+                self.money += 200
+                print(f"{self.name} passed Go and collects $200.")
+        if result[0] != result[1]:
+            self.is_turn = False
+        return result
+    
+    def take_turn(self):
+        self.is_turn = True
+        turn_count = 0
+        # Rolls dice and repeats if doubles are achieved, unless it's done three times in a row
+        while self.is_turn or turn_count < 3:
+            turn_count += 1
+            self.update_position()
+            if self.is_turn and turn_count < 3:
+                print(f"{self.name} rolled doubles! {self.name} goes again!")
+            elif turn_count >= 3:
+                print(f"{self.name} rolled doubles three times in a row! {self.name} goes to jail!")
+                self.position = 10
+                self.in_jail = True
+
+
+class Property:
+    def __init__(self, property_name, property_cost, property_house_cost, property_rent0, property_rent1, property_rent2, property_rent3, property_rent4, property_rent5, property_mortgage):
+        self.name = property_name
+        self.cost = property_cost
+        self.house_cost = property_house_cost
+        self.houses = 0
+        self.rent0 = property_rent0
+        self.rent1 = property_rent1
+        self.rent2 = property_rent2
+        self.rent3 = property_rent3
+        self.rent4 = property_rent4
+        self.rent5 = property_rent5
+        self.mortgage = property_mortgage
+    
+    def __repr__(self):
+        return f"{self.name}: ${self.cost}. \nRent ${self.rent0}. \nWith 1 House: ${self.rent1}. \nWith 2 Houses: ${self.rent2}. \nWith 3 Houses: ${self.rent3}. \nWith 4 Houses: ${self.rent4}. \nWith Hotel: ${self.rent5}. \n\nMortgage Value: ${self.mortgage}. \nHouses cost ${self.house_cost} each. \nHotels, ${self.house_cost} plus 4 houses. \nIf a player owns ALL the lots in any Color-Group, the rent is Doubled on Unimproved Lots in that group."
+    
+    def buy_property(self, player):
+        if player.money < self.cost:
+            print(f"You don't have enough money to afford this property.")
+            return
+        else:
+            player.properties.append(self)
```

Comment:
A question to consider when making class-based design: which class handles the logic? Here we have the `Property` class modifying a property on the player; if we ever change the `Player` class to store properties differently, we also have to change the `Property` class to handle those changes.

There's no objectively right design, but as you work on this you'll start to notice how [tightly coupled](https://en.wikipedia.org/wiki/Coupling_(computer_programming)) classes can become, which often reduces the utility of them.

Reply:
Hmm, I see. The course in this case asked me to make the classes interact with each other so that's what I was going for here.